src/Lean/Elab/Tactic/Induction.lean

```diff
@@ -583,16 +596,16 @@ def elabCasesTargets (targets : Array Syntax) : TacticM (Array Expr) :=
             j := j+1
           else
             result := result.push arg.expr
-        return (result, [mvarId])
+        return ((result, hIdents.zip fvarIdsNew[j:]), [mvarId])
```

Comment:
The approach is good but this indexing looks dubious to me, what exactly is happening here? How can this be correct if `shouldGeneralizeTarget` is true for at least one arg?

Reply:
The return value of `generalize` is a list of fvarIds with exactly one entry for each `x` variable provided, followed by the `h` variables in order, skipping the ones that are `none`. The `filterM` line trims some of the variables, but because of the `<||> arg.hName?.isSome` part every `h` variable is preserved.

We call `generalize` with the filtered list, so it now has `fvarIdsNew` containing the selected `x` variables and all of the `h` variables. The variable `j` counts the `x` variables which have not been filtered out, so at the end of the loop it has enumerated all the `x` variables and what remains is the list of all the `h` variables, which exactly match what we first put in the `hIdents` array.

(Yes, this indexing design is horribly convoluted, but I don't feel like refactoring `generalize` right now. IMO a better design would be to return a list of `FVar x Option FVarId` pairs matching the input `GeneralizeArg`s.)

Reply:
Ohh, I had no idea about the makeup of the `generalize` result. Then the indexing makes sense, thanks. Maybe a short comment to that effect on this line?